\documentclass{article}
\usepackage{amsmath}
\usepackage[margin=3cm]{geometry}
\usepackage[hidelinks]{hyperref}

\makeatletter 
\renewcommand{\fnum@figure}{Supplementary \figurename~\thefigure}
\renewcommand{\fnum@table}{Supplementary \tablename~\thetable}
\makeatother

\usepackage[labelfont=bf]{caption}
\usepackage{subcaption}
\usepackage{graphicx}
\newcommand{\code}[1]{\texttt{#1}}

\usepackage{color}
\newcommand{\revised}[1]{\textcolor{red}{#1}}

\begin{document}

\begin{titlepage}
\vspace*{3cm}
\begin{center}


{\LARGE
Distinguishing cells from empty droplets in droplet-based single-cell RNA sequencing data
\par}

\vspace{0.75cm}

{\Large
    \textsc{Supplementary Materials}
\par
}
\vspace{0.75cm}

\large
by


\vspace{0.75cm}
Aaron T. L. Lun$^1$,
and others

\vspace{1cm}
\begin{minipage}{0.9\textwidth}
\begin{flushleft}
$^1$Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom \\[6pt]
$^2$Blah blah blah \\[6pt]
\end{flushleft}
\end{minipage}

\vspace{1.5cm}
{\large \today{}}

\vspace*{\fill}
\end{center}
\end{titlepage}

\section*{Motivating the choice of the total count threshold $T$}
The threshold $T$ should be chosen so that cell-containing droplets are not used to estimate the ambient profile.
Otherwise, our ambient profile estimate will be distorted and we will have less power to discriminate between cells and empty droplets.
To check whether this occurs in real data, we calculated a $p$-value against the ambient null hypothesis for each barcode $b \in \mathcal{G}$, 
i.e., with $t_b \le T$ where $T = 100$ by default.
If many cell-containing droplets were present in $\mathcal{G}$, we should observe an enrichment of low $p$-values.
These low $p$-values either correspond to the cell-containing droplets themselves,
or to genuinely empty droplets that no longer fit to our distorted estimate of the ambient profile.

We do not observe any enrichment for low $p$-values in any tested dataset (Supplementary Figure~\ref{fig:negative}). 
This is true even when the $p$-values were weighted by $t_b$ to represent the contribution of each barcode to the ambient profile.
If cell-containing droplets with low $p$-values and large $t_b$ were present, enrichment of low $p$-values should be more prominent in the weighted distribution.
However, this is not the case, indicating that distortions of the ambient profile due to cell-containing droplets are not an issue with $T=100$.
Instead, there is an enrichment at a $p$-value of 0.5, indicating that we were overstating the variance of the deviances in our model.
This manifests as conservativeness -- most likely due to the fact that we were not conditioning on the total UMI count for each barcode -- and is acceptable.

\newpage
\begin{figure}[btp]
    \begin{center}
        \includegraphics[width=0.4\textwidth]{../simulations/pics-negcheck/hist_293t.pdf}
        \includegraphics[width=0.4\textwidth]{../simulations/pics-negcheck/hist_jurkat.pdf}
        \includegraphics[width=0.4\textwidth]{../simulations/pics-negcheck/hist_neuron_9k.pdf}
        \includegraphics[width=0.4\textwidth]{../simulations/pics-negcheck/hist_neurons_900.pdf}
        \includegraphics[width=0.4\textwidth]{../simulations/pics-negcheck/hist_pbmc4k.pdf}
        \includegraphics[width=0.4\textwidth]{../simulations/pics-negcheck/hist_t_4k.pdf}
    \end{center}
    \caption{Distribution of $p$-values for all barcodes with total UMI counts less than or equal to the threshold $T=100$.
        Each plot corresponds to a dataset in Supplementary Table~\ref{tab:datasets} where the $p$-value represents the deviation from the ambient profile.
        Bars represent the number of barcodes for each $p$-value interval, while the dotted line represents the contribution of all barcodes in that interval to the ambient profile.
        The contribution from each interval is calculated as the sum of counts across all of its barcodes and is shown as a percentage of the total count in the ambient profile.
    }
    \label{fig:negative}
\end{figure}

\begin{figure}[btp]
    \begin{center}
        \includegraphics[width=\textwidth]{../simulations/pics-sim/pbmc4k_2000_2000.pdf}
    \end{center}
\caption{Total count against the rank for each barcode in a simulation based on the PBMC dataset with $G_1=G_2=2000$.
Plots are shown for all barcodes, barcodes corresponding to empty droplets, and barcodes corresponding to large or small cells.
Ranks are calculated from the entire set of barcodes in all plots, for ease of comparison between plots.
All axes are on a log-scale.}
\end{figure}

\begin{table}[btp]
\caption{Summary of the datasets used to assess the various cell detection methods.
    All datasets were obtained from the 10X Genomics website.
The organism, cell type, estimated number of cells from CellRanger and the version of the CellRanger software used is shown for each dataset.}
\begin{center}
\begin{tabular}{l l l l r}
\hline
\textbf{Name} & \textbf{Organism} & \textbf{Cell type} & \textbf{Number} & \textbf{Version} \\
\hline
\code{293t}   & Human & 293T cell line & 2800 & 1.1.0 \\
\code{jurkat} & Human & Jurkat cell line & 3200 & 1.1.0 \\
\code{neuron\_9k} & Mouse & Brain cells & 9128 & 2.1.0 \\
\code{neurons\_900} & Mouse & Brain cells & 931 & 2.1.0 \\
\code{pbmc4k} & Human & PBMCs & 4340 & 2.1.0 \\
\code{t\_4k} & Human & Pan T cells & 4538 & 2.1.0 \\
\hline
\end{tabular}
\end{center}
\label{tab:datasets}
\end{table}

\end{document}
