\documentclass{article}
\usepackage{amsmath}
\usepackage[margin=3cm]{geometry}
\usepackage[hidelinks]{hyperref}

\makeatletter 
\renewcommand{\fnum@figure}{Supplementary \figurename~\thefigure}
\renewcommand{\fnum@table}{Supplementary \tablename~\thetable}
\makeatother

\usepackage[labelfont=bf]{caption}
\usepackage{subcaption}
\usepackage{graphicx}
\newcommand{\code}[1]{\texttt{#1}}

\usepackage{color}
\newcommand{\revised}[1]{\textcolor{red}{#1}}

\newcommand{\PlotSim}[2]{%
\begin{figure}[btp]
    \begin{center}
        \begin{subfigure}{0.8\textwidth}
        \includegraphics[width=\textwidth,trim=0mm 15mm 0mm 0mm,clip]{../simulations/results-sim/pics/#1-recall.pdf}
        \caption{}
        \end{subfigure}
        \begin{subfigure}{0.8\textwidth}
        \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 0mm,clip]{../simulations/results-sim/pics/#1-fdr.pdf}
        \caption{}
        \end{subfigure}
    \end{center}
\caption{Results for the simulations based on the #2 dataset, using either EmptyDrops or the default CellRanger method for detecting cell-containing droplets.
Simulation scenarios are labelled as $G_1/G_2$ where $G_1$ and $G_2$ are the number of barcodes in the group of large and small cells, respectively.
(a) The recall for each method, defined as the proportion of detected cells from each group.
EmptyDrops was used with a FDR threshold of 0.1\%.
(b) The observed FDR in the set of libraries detected by EmptyDrops at a range of nominal FDR thresholds (dotted lines), defined as the proportion of detected droplets that are empty.
In both plots, each point represents the result of one simulation iteration, the bar represents the mean value across 10 iterations and the error bars represent the standard error of the mean.
}
\end{figure}}

\newcommand{\PlotReal}[2]{%
\begin{figure}[btp]
    \begin{center}
    \begin{subfigure}{0.49\textwidth}
        \includegraphics[width=\textwidth,trim=0mm 0mm 0mm 20mm,clip]{../real/pics-real/rank_#1.pdf}
        \caption{}
    \end{subfigure}
    \begin{subfigure}{0.49\textwidth}
        \includegraphics[width=\textwidth,trim=0mm 0mm 0mm 20mm,clip]{../real/pics-real/dev_#1.pdf}
        \caption{}
    \end{subfigure}\\[0.1in]
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\textwidth,trim=0mm 0mm 0mm 5mm,clip]{../real/pics-real/intersect_#1.pdf}
        \caption{}
    \end{subfigure}
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\textwidth,trim=0mm 0mm 0mm 20mm,clip]{../real/pics-real/kept_#1.pdf}
        \caption{}
    \end{subfigure}
\end{center}
    \caption{Results of applying EmptyDrops and the other cell detection methods to the #2 dataset.
        (a) A barcode rank plot showing the fitted spline used for knee point detection in EmptyDrops. 
        The detected knee and inflection points are also shown.
        (b) The negative log-probability for each barcode in the multinomial model of EmptyDrops, plotted against the total count.
        Barcodes detected as putative cell-containing droplets at a FDR of 1\% are labelled in red.
        Only barcodes with $t_b > T$ are shown.
        (c) An UpSet plot of the barcodes detected by each combination of methods (vertical bars).
        Horizontal bars represent the number of barcodes detected by each method.
        (d) Histogram outlines of the log-total count for barcodes detected by each method.}
\end{figure}}

\begin{document}

\begin{titlepage}
\vspace*{3cm}
\begin{center}


{\LARGE
Distinguishing cells from empty droplets in droplet-based single-cell RNA sequencing data
\par}

\vspace{0.75cm}

{\Large
    \textsc{Supplementary Materials}
\par
}
\vspace{0.75cm}

\large
by


\vspace{0.75cm}
\begin{minipage}{0.9\textwidth}
\begin{flushleft}
Aaron T. L. Lun\textsuperscript{1,*,\dag{}},
Samantha Riesenfeld\textsuperscript{2,*},
Tallulah Andrews\textsuperscript{3,*},
The Phuong Dao\textsuperscript{4,*},
Tomas Gomes\textsuperscript{3,*},
participants in the 1\textsuperscript{st} Human Cell Atlas Jamboree\textsuperscript{\ddag},
John C. Marioni\textsuperscript{1,3,5,\#}
\end{flushleft}
\end{minipage}

\vspace{1cm}
\begin{minipage}{0.9\textwidth}
\begin{flushleft}
$^1$Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom \\[6pt]
$^2$Klarman Cell Observatory, Broad Institute of MIT and Harvard, Cambridge, Massachusetts, USA \\[6pt]
$^3$Wellcome Trust Sanger Institute, Wellcome Genome Campus, Hinxton, Cambridge CB10 1SA, United Kingdom \\[6pt]
$^4$Program for Computational and Systems Biology, Sloan Kettering Institute, Memorial Sloan Kettering Cancer Center, New York, NY \\[6pt]
$^5$EMBL European Bioinformatics Institute, Wellcome Genome Campus, Hinxton, Cambridge CB10 1SD, United Kingdom
\end{flushleft}
* These authors contributed equally to this work. \\
\dag{} Email: aaron.lun@cruk.cam.ac.uk \\
\ddag{} The full list of participants is provided in Supplementary Table 1. \\
\# Email: marioni@ebi.ac.uk
\end{minipage}

\vspace{1.5cm}
{\large \today{}}

\vspace*{\fill}
\end{center}
\end{titlepage}

\section{Motivating the choice of the total count threshold $T$}
The threshold $T$ should be chosen so that cell-containing droplets are not used to estimate the ambient profile.
Otherwise, our estimate will be distorted and we will have less power to discriminate between cells and empty droplets.
To check if this occurs in real data, we calculated a $p$-value against the ambient null hypothesis for each barcode $b \in \mathcal{G}$, 
i.e., with $t_b \le T$ where $T = 100$ by default.
Ideally, we should observe a uniform distribution of $p$-values under the assumption that all barcodes in $\mathcal{G}$ are genuinely empty.
However, if cell-containing droplets are present in $\mathcal{G}$, we should observe an enrichment of low $p$-values.
These low $p$-values either correspond to the cell-containing droplets themselves,
or to genuinely empty droplets that no longer fit to our distorted estimate of the ambient profile.

In most datasets, we observed a minor enrichment of low $p$-values in an otherwise uniform distribution (Supplementary Figure~\ref{fig:negative}). 
This demonstrates that our testing procedure mostly holds its size and suggests that distortions of the ambient profile due to cell-containing droplets are not a major issue when using $T=100$.
The exception is the \code{neurons\_900} dataset where we see a clear enrichment of very low and very large $p$-values.
Note that  this effect is not consistent with distortion of the ambient profile, which should only result in a skew towards low $p$-values.
Rather, we hypothesize that it is driven by violations of the independent sampling assumption (e.g., if transcript molecules are complexed and sampled together into droplets).
Positive correlations between molecules would increase the probability of obtaining droplets that are very similar or very different to the ambient profile.

\section{Efficient calculation of the Monte Carlo $p$-value}
The naive approach to computing the Monte Carlo $p$-value for a barcode would be to directly sample $R$ count vectors from a multinomial distribution,
compute the likelihood $L'_{bi}$ for each count vector, and count the number of vectors with $L'_{bi} \le L_{b}$.
For $N$ genes, we would need $RN$ sampling operations to obtain the count vectors and the same number of multiplication operations to compute the likelihoods.
For a dataset containing $B$ barcodes, this would amount to a total of $RNB$ operations.

When dealing with multiple barcodes, we can improve efficiency by using pre-existing results for barcodes with lower $t_b$.
Within a simulation iteration $i$, we sample a count vector and calculate the likelihood $L'_{b_1i}$ for the barcode $b_1$ with the lowest total count.
For barcode $b_2$ with the total $t_{b_2}=t_{b_1}+1$, we update the count vector by sampling a new gene $k$ from the multinomial distribution with a size of 1 and probabilities equal to our ambient proportions.
Similarly, we update the likelihood 
\[
    L'_{b_2i} = L'_{b_1i} \frac{t_{b_2} \tilde{p}_k}{y'_{ki}} \;,
\]
where $y'_{ki}$ is the frequency of $k$ in the updated count vector.
(Iterative application of these updates can be used in cases where $t_{b_2} > t_{b_1}+1$.)
The updated $L'_{b_2i}$ can be directly compared to $L_{b_2}$, and repeating this procedure for $R$ iterations allows us to compute the $p$-value for each $b$.
This approach only requires $R \max\{t_b\}$ sampling and multiplication operations, 
and for droplet-based data, the maximum $t_b$ across all barcodes in a dataset is much lower than $NB$.
The former typically has values of 5000-10000, while the number of detected genes $N$ is around 3000-5000 and $B$ for all $t_b > T$ is often over 1000.
This results in a considerable reduction in computational work compared to the naive approach.

Our algorithm can be viewed as an extension of a more simple speed-up where the set of likelihoods from all iterations are re-used for all barcodes with the same $t_b$.
As a result, though, the Monte Carlo $p$-values are not strictly independent across barcodes.
Instead, they are positively correlated as the same sampling choices are effectively shared between barcodes.
This poses a theoretical problem for multiple testing correction with the Benjamini-Hochberg (BH) method, which relies on independent $p$-values.
In practice, this is not a major issue as the BH method is robust to dependencies \cite{reiner2003identifying,kim2008effects}.
Moreover, increasing $R$ reduces the imprecision of the $p$-values caused by random sampling.
This means that the effect of correlated estimation errors between different $p$-values will become negligible at large $R$.
We use $R=10000$ by default, though this can be increased if the lower bound on the $p$-values \cite{phipson2010permutation} is greater than the threshold for significance after the Benjamini-Hochberg correction.

\bibliography{ref.bib}
\bibliographystyle{unsrt}

\newpage
\begin{figure}[btp]
    \begin{center}
        \includegraphics[width=0.4\textwidth]{../real/pics-negcheck/hist_293t.pdf}
        \includegraphics[width=0.4\textwidth]{../real/pics-negcheck/hist_jurkat.pdf}
        \includegraphics[width=0.4\textwidth]{../real/pics-negcheck/hist_neuron_9k.pdf}
        \includegraphics[width=0.4\textwidth]{../real/pics-negcheck/hist_neurons_900.pdf}
        \includegraphics[width=0.4\textwidth]{../real/pics-negcheck/hist_pbmc4k.pdf}
        \includegraphics[width=0.4\textwidth]{../real/pics-negcheck/hist_t_4k.pdf}
    \end{center}
    \caption{Histograms of $p$-values for all barcodes with total UMI counts less than or equal to the threshold $T=100$.
        Each plot corresponds to a dataset in Supplementary Table~\ref{tab:datasets} where the $p$-value represents the deviation from the ambient profile.
    }
    \label{fig:negative}
\end{figure}

\begin{figure}[btp]
    \begin{center}
        \includegraphics[width=\textwidth]{../simulations/pics-sim/pbmc4k_2000_2000.pdf}
    \end{center}
\caption{Total count against the rank for each barcode in a simulation based on the PBMC dataset with $G_1=G_2=2000$.
Plots are shown for all barcodes, barcodes corresponding to empty droplets, and barcodes corresponding to large or small cells.
Ranks are calculated from the entire set of barcodes in all plots, for ease of comparison between plots.
All axes are on a log-scale.}
\end{figure}

\begin{table}[btp]
\caption{Summary of the datasets used to assess the various cell detection methods.
    All datasets were obtained from the 10X Genomics website.
The organism, cell type, expected number of cells and the version of the CellRanger software used is shown for each dataset.}
\begin{center}
\begin{tabular}{l l l r r}
\hline
\textbf{Name} & \textbf{Organism} & \textbf{Cell type} & \textbf{Number} & \textbf{Version} \\
\hline
\code{pbmc4k} & Human & PBMCs & 4000 & 2.1.0 \\
\code{293t}   & Human & 293T cell line & 2800 & 1.1.0 \\
\code{jurkat} & Human & Jurkat cell line & 3200 & 1.1.0 \\
\code{neuron\_9k} & Mouse & Brain cells & 9000 & 2.1.0 \\
\code{neurons\_900} & Mouse & Brain cells & 900 & 2.1.0 \\
\code{t\_4k} & Human & Pan T cells & 4000 & 2.1.0 \\
\hline
\end{tabular}
\end{center}
\label{tab:datasets}
\end{table}

\PlotSim{293t}{293T cell line}
\PlotSim{jurkat}{Jurkat cell line}
\PlotSim{neuron_9k}{9K brain cell}
\PlotSim{neurons_900}{900 brain cell}
\PlotSim{t_4k}{pan T cell}

\PlotReal{293t}{293T cell line}
\PlotReal{jurkat}{Jurkat cell line}
\PlotReal{neuron_9k}{9K brain cell}
\PlotReal{neurons_900}{900 brain cell}
\PlotReal{t_4k}{pan T cell}


\begin{figure}[btp]
    \begin{subfigure}{0.49\textwidth}
        \includegraphics[width=\textwidth]{../analysis/neurons_900/pics/by_detection.pdf}
        \caption{}
    \end{subfigure}
    \begin{subfigure}{0.49\textwidth}
        \includegraphics[width=\textwidth]{../analysis/neurons_900/pics/by_interneuron.pdf}
        \caption{}
    \end{subfigure} \\[0.05in]
    \begin{subfigure}{0.49\textwidth}
        \includegraphics[width=\textwidth]{../analysis/neurons_900/pics/by_ribo.pdf}
        \caption{}
    \end{subfigure}
    \begin{subfigure}{0.49\textwidth}
        \includegraphics[width=\textwidth]{../analysis/neurons_900/pics/by_mito.pdf}
        \caption{}
    \end{subfigure}
    \caption{$t$-SNE plots for the 900 brain cell dataset, constructed using the first 100 principal components of the normalized log-expression matrix.
        Each point represents a cell that is coloured by (a) detection with EmptyDrops alone or both CellRanger and EmptyDrops, 
        (b) expression of interneuron markers \textit{Gad1}, \textit{Gad2} and \textit{Sla6c1}, (c) expression of ribosomal protein genes or (d) the proportion of counts assigned to mitochondrial genes.
        Expression in each cell was quantified as the sum of the normalized log-expression values across all genes in the relevant set.
        Mitochondrial proportions are capped at 20\% to improve visibility.     
        The arrow marks the putative interneuron population.
    }
    \label{fig:realtsne}
\end{figure}


\end{document}
